google-cloud-spanner/src/main/java/com/google/cloud/spanner/Interval.java

```diff
@@ -0,0 +1,316 @@
+  public static final long MILLIS_PER_SECOND = 1000;
+  public static final long MICROS_PER_MILLI = 1000;
+  public static final long NANOS_PER_MICRO = 1000;
+  public static final long MICROS_PER_SECOND = MICROS_PER_MILLI * MILLIS_PER_SECOND;
+  public static final long MICROS_PER_MINUTE = SECONDS_PER_MINUTE * MICROS_PER_SECOND;
+  public static final long MICROS_PER_HOUR = SECONDS_PER_HOUR * MICROS_PER_SECOND;
+  public static final long MICROS_PER_DAY = HOURS_PER_DAY * MICROS_PER_HOUR;
+  public static final long MICROS_PER_MONTH = DAYS_PER_MONTH * MICROS_PER_DAY;
+  public static final BigInteger NANOS_PER_SECOND =
+      BigInteger.valueOf(MICROS_PER_SECOND * NANOS_PER_MICRO);
+  public static final BigInteger NANOS_PER_MINUTE =
+      BigInteger.valueOf(MICROS_PER_MINUTE * NANOS_PER_MICRO);
+  public static final BigInteger NANOS_PER_HOUR =
+      BigInteger.valueOf(MICROS_PER_HOUR * NANOS_PER_MICRO);
+
+  private static final Pattern INTERVAL_PATTERN =
+      Pattern.compile(
+          "^P(?!$)(-?\\d+Y)?(-?\\d+M)?(-?\\d+D)?(T(?=-?.?\\d)(-?\\d+H)?(-?\\d+M)?(-?((\\d+(\\.\\d{1,9})?)|(\\.\\d{1,9}))S)?)?$");
+
+  /** Returns the months component of the interval. */
+  public abstract int months();
+
+  /** Returns the days component of the interval. */
+  public abstract int days();
+
+  /** Returns the microseconds component of the interval. */
+  public abstract long micros();
+
+  /** Returns the nanoFractions component of the interval. */
+  public abstract short nanoFractions();
+
+  public static Builder builder() {
+    return new AutoValue_Interval.Builder();
+  }
+
+  /** Returns the nanoseconds component of the interval. */
+  public BigInteger nanos() {
+    return BigInteger.valueOf(micros())
+        .multiply(BigInteger.valueOf(NANOS_PER_MICRO))
+        .add(BigInteger.valueOf(nanoFractions()));
+  }
+
+  /** Returns the total microseconds represented by the interval. */
+  public long getAsMicros() {
+    return months() * MICROS_PER_MONTH + days() * MICROS_PER_DAY + micros();
+  }
+
+  /** Returns the total nanoseconds represented by the interval. */
+  public BigInteger getAsNanos() {
+    return BigInteger.valueOf(getAsMicros())
+        .multiply(BigInteger.valueOf(NANOS_PER_MICRO))
+        .add(BigInteger.valueOf(nanoFractions()));
+  }
+
+  /** Creates an interval with specified number of months. */
+  public static Interval ofMonths(int months) {
+    return builder().setMonths(months).setDays(0).setMicros(0).setNanoFractions((short) 0).build();
+  }
+
+  /** Creates an interval with specified number of days. */
+  public static Interval ofDays(int days) {
+    return builder().setMonths(0).setDays(days).setMicros(0).setNanoFractions((short) 0).build();
+  }
+
+  /** Creates an interval with specified number of seconds. */
+  public static Interval ofSeconds(long seconds) {
+    return builder()
+        .setMonths(0)
+        .setDays(0)
+        .setMicros(seconds * MICROS_PER_SECOND)
```

Comment:
We should either document what happens if you specify a value for seconds that will make this overflow, or check it and return an error.